### PR TITLE
ENH: Add nullable dtypes to read_csv

### LINKED ISSUE #36712: BUG: `pd.read_csv` not preserving Nullable types
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
In[2]: import pandas as pd
In[3]: s = pd.Series([1, 2, None, 3], dtype=pd.UInt64Dtype())
In[4]: s
Out[4]: 
0       1
1       2
2    <NA>
3       3
dtype: UInt64
In[5]: df = pd.DataFrame({"a":s})
In[6]: df
Out[6]: 
      a
0     1
1     2
2  <NA>
3     3
In[7]: df.to_csv('temp.csv')
In[8]: pd.read_csv('temp.csv')
Out[8]: 
   Unnamed: 0    a
0           0  1.0
1           1  2.0
2           2  NaN
3           3  3.0

```

#### Problem description

While reading csv, pandas should likely read in the above dataframe column as a nullable integer dtype instead of converting to floating data type. 

#### Expected Output

```python
pd.read_csv('temp.csv')
Out[8]: 
   Unnamed: 0  a
0           0  1
1           1  2
2           2  <NA>
3           3  3
```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2a7d3326dee660824a8433ffd01065f8ac37f7d6
python           : 3.7.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-48-generic
Version          : #52-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.1.2
numpy            : 1.19.1
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 49.6.0.post20200917
Cython           : 0.29.21
pytest           : 6.0.2
hypothesis       : 5.28.0
sphinx           : 3.2.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 0.8.2
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 1.0.1
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.51.2

</details>

### PR DESCRIPTION
**Pull request by lithomas1**
- [x] closes #36712 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/dtypes/test_inference.py
Comment (gfyoung) [inline pandas/tests/dtypes/test_inference.py:622]: Do we have/need a similar test for nullable boolean?

**Comment by gfyoung** [comment]
As a whole looks pretty good, though with the growing number of null-related parameters, I think we will at some point need to condense into one "super parameter" (e.g. `dict` of some sorts).

File: pandas/_libs/lib.pyx
Comment (jbrockmendel) [inline pandas/_libs/lib.pyx:2009]: can this be done at a higher level (i.e. not in the cython code)?

Comment (lithomas1) [reply pandas/_libs/lib.pyx:2009]: That would probably require a substantial refactoring of the Python Parser code, which calls this directly. 

FWIW, maybe_convert_objects also has a convert_to_nullable_integer param.

**Comment by jorisvandenbossche** [comment]
@lithomas1 Cool! Thanks for working on this.

We now also have a nullable float dtype nowadays. I know that this is less useful at this point (since float already supports missing values, in contrast to integer/boolean), but I think that a `use_nullable_dtypes` keyword should probably also result in nullable float.

> As a whole looks pretty good, though with the growing number of null-related parameters, I think we will at some point need to condense into one "super parameter" (e.g. dict of some sorts).

@gfyoung I don't think this keyword is very related to the other read_csv null-related keywords? (or which ones were you thinking about?) And the `use_nullable_dtypes` keyword is also consistent with other IO methods (#29752, #31242)

File: pandas/tests/io/parser/test_na_values.py
Comment (jorisvandenbossche) [inline pandas/tests/io/parser/test_na_values.py:161]: We should probably use "boolean" dtype if `use_nullable_dtypes=True` regardless of whether NA's where actually present or not?

**Comment by gfyoung** [comment]
> @gfyoung I don't think this keyword is very related to the other read_csv null-related keywords? (or which ones were you thinking about?) And the use_nullable_dtypes keyword is also consistent with other IO methods (#29752, #31242)

@jorisvandenbossche: They are related broadly speaking, but the consistency argument is fair as well.

I was just saying it's something to consider in the future, especially if we continue to add more arguments to an already long signature for `read_csv` (and the other I/O methods).

File: pandas/tests/io/parser/test_na_values.py
Comment (lithomas1) [reply pandas/tests/io/parser/test_na_values.py:161]: All NAs has unknown dtype, so we follow numpy and return float64. Currently, there has to be an integer or boolean value and NAs in order to return a nullable EA type, so a column of a float or bool will just return the numpy types.

**Review by jreback — CHANGES_REQUESTED** [review]
can you do a pre-cursor PR which does the lower level changes (e.g. to maybe_convert_*) and fully unit tests those. then we can layer to reader changes in this PR.

**Comment by lithomas1** [comment]
@jreback I think I will first try to get this feature-complete by adding in support for nullable float and string dtypes. Then I can spit out the inferencing changes. Marking this as draft in the meantime.

**Comment by lithomas1** [comment]
Currently progress blocked on StringArray support by #40839

**Comment by jorisvandenbossche** [comment]
@lithomas1 I think that the resulting dtype when using `use_nullable_dtypes=True` should not depend on whether NAs are present or not (IIUC the current state of the PR will eg only use nullable Int64 dtype if there is at least one missing value in the column, and otherwise return plain int64 dtype?) 
Having it depend on the presence if missing values introduces "value-dependent" behaviour, which we generally try to avoid. The nullable vs non-nullable dtype *can* behave somewhat differently, and so IMO it's better to return consistent dtypes. For example, assume you have a bunch of CSV files with the same structure (columns and types), then ideally you get the same inferred dtypes for each file (except for the corner case of all-missing data, of course).

**Comment by jbrockmendel** [comment]
the downside of that is it locks the user in to non-consolidated columns, with associated perf penalties.

**Comment by lithomas1** [comment]
@jorisvandenbossche That is correct, the nullable dtypes are not used unless null values appear in the data. For everything except StringArray/object, I think the only behavior difference would be more memory usage and less supported methods, if we have a nullable array and a regular array without null values. 
For StringArray vs object dtype for a column with all strings and no null values, I guess I'm fine with returning StringArray.

File: pandas/_libs/parsers.pyx
Comment (simonjayhawkins) [inline pandas/_libs/parsers.pyx:1370]: The constructors for StringArray and ArrowStringArray are incompatible. To allow the global default for StringDtype backend storage (pyarrow or python) to be used here in the future (after #39908 is merged), we should probably call StringDtype.construct_array_type._from_sequence.

**Comment by jreback** [comment]
@lithomas1 status of this?

**Comment by jreback** [comment]
@lithomas1 this might be easier these days?

**Comment by lithomas1** [comment]
Still stuck by at least #45168.

**Comment by jreback** [comment]
@lithomas1 status of this?

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. Feel free to reopen when you have the time to revisit. Closing to clear out the queue.

**Pull request closed without merging**
